packages/server/test/websockets.test.ts

```diff
@@ -0,0 +1,169 @@
+    trpc.router().query('hello', {
+      resolve() {
+        return 'there';
+      },
+    }),
+    {
+      client({ wssUrl }) {
+        ws = createWebSocketClient({ url: wssUrl });
+        return {
+          links: [wsLink({ client: ws })],
+        };
+      },
+    },
+  );
+  const res = await client.query('hello');
+  expect(res).toBe('there');
+
+  close();
+  ws.close();
+});
+
+test('subscriptionOnce()', async () => {
+  const ee = new EventEmitter();
+  type Message = {
+    id: string;
+  };
+  let ws: any = null;
+
+  const { client, close } = routerToServerAndClient(
+    trpc.router().subscription('onMessage', {
+      input: z.string().optional(),
+      resolve() {
+        ee.emit('server:connect');
+        return new trpc.Subscription<Message>({
+          start(emit) {
+            const onMessage = (data: Message) => {
+              emit.data(data);
+            };
+            ee.on('server:msg', onMessage);
+            return () => ee.off('server:msg', onMessage);
+          },
+        });
+      },
+    }),
+    {
+      client({ wssUrl }) {
+        ws = createWebSocketClient({ url: wssUrl });
+        return {
+          links: [wsLink({ client: ws })],
+        };
+      },
+    },
+  );
+  ee.once('server:connect', () => {
+    setImmediate(() => {
+      ee.emit('server:msg', {
+        id: '1',
+      });
+    });
+  });
+  const msgs = await client.subscriptionOnce('onMessage', '');
+
+  expectTypeOf(msgs).toMatchTypeOf<Message[]>();
+  expect(msgs).toMatchInlineSnapshot(`
+    Object {
+      "id": "1",
+    }
+  `);
+
+  await waitFor(() => {
+    expect(ee.listenerCount('server:msg')).toBe(0);
+    expect(ee.listenerCount('server:error')).toBe(0);
+  });
+  close();
+  ws.close();
+});
+
+test('$subscription()', async () => {
+  const ee = new EventEmitter();
+  type Message = {
+    id: string;
+  };
+  let ws: TRPCWebSocketClient = null as any;
+  const { client, close } = routerToServerAndClient(
+    trpc.router().subscription('onMessage', {
+      input: z.string().optional(),
+      resolve() {
+        ee.emit('server:connect');
+        return new trpc.Subscription<Message>({
+          start(emit) {
+            const onMessage = (data: Message) => {
+              emit.data(data);
+            };
+            ee.on('server:msg', onMessage);
+            return () => ee.off('server:msg', onMessage);
+          },
+        });
+      },
+    }),
+    {
+      client({ wssUrl }) {
+        ws = createWebSocketClient({ url: wssUrl });
+        return {
+          links: [wsLink({ client: ws })],
+        };
+      },
+    },
+  );
+  ee.once('server:connect', () => {
+    setImmediate(() => {
+      ee.emit('server:msg', {
+        id: '1',
+      });
+      ee.emit('server:msg', {
+        id: '1',
+      });
+    });
+  });
+  const callback = jest.fn();
+  const $sub = client.$subscription('onMessage');
+
+  $sub.subscribe({ onNext: callback });
```

Comment:
The subscriptions is in this format and you have to call `$sub.done()` to end the subscription. This allows multiple subscribers, but maybe that's over-doing it.

Could be nicer with this format:

```ts
const unsubscribe = client.$subscription('onMessage', {
  onNext(data: TOutput) {},
  onError(err: TRPCClientError) {},
});

unsubscribe();
```

Thoughts? 

---


Secondly - error handling - there are some errors that would be "fatal" and some that at "transient". I.e. one pull of the subscription in the backend could return an error which wouldn't cause the link to broken, but then there's other type of errors that could force-quit the subscription... unsure how to deal with the different cases. 


---

cc @cyrus-za @anthonyshort

Reply:
I really like that format. Not sure if I prefer `onData` over `onNext`

---

For error handling, inversion of control is what seems most "appropiate" for me, so the developer should handle the different types of errors instead of trpc handling them. What I'm thinking is that whenever any error occurs, that should trigger the `onError` callback, maybe this callback's error could also have a `code`? For example if the connection closes abnormally, the code would be `1006`, all the websocket error codes are listed here: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#properties. 

This way, the developer using trpc should handle the different type of errors based on whether they have a code or not:
```ts
const unsubscribe = client.$subscription('onMessage', {
  onNext(data: TOutput) {},
  onError(err: TRPCClientError, code?: number) {
    if (code) {
      // Do something...
    }
    
    // Do something else based on `err`
  },
});

unsubscribe();
```
Thoughts?

Reply:
I have changed to your suggestion, @gmencz. What do you think now?

Still unsure how to handle errors nicely...

Reply:
variants:

- intermittent error - one pull messed up but next `onNext()` could work
- fatal error for the specific subscription - an error led to subscription being cancelled
- error because server shut down
  - potentially because of a new version on the server - would be nice to handle gracefully
  - fails but potentially reconnect
- ..